File: app/src/main/java/org/wikipedia/dataclient/mwapi/MwQueryResult.java
package org.wikipedia.dataclient.mwapi;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.annotations.SerializedName;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.dataclient.WikiSite;
import org.wikipedia.dataclient.page.Protection;
import org.wikipedia.json.PostProcessingTypeAdapter;
import org.wikipedia.notifications.Notification;
import org.wikipedia.page.PageTitle;
import org.wikipedia.settings.SiteInfo;
import org.wikipedia.util.DateUtil;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Map;

@SuppressWarnings("unused")
public class MwQueryResult implements PostProcessingTypeAdapter.PostProcessable {
    @Nullable private List<MwQueryPage> pages;
    @Nullable private List<Redirect> redirects;
    @Nullable private List<ConvertedTitle> converted;
    @SerializedName("userinfo") private UserInfo userInfo;
    @Nullable private List<ListUserResponse> users;
    @Nullable private Tokens tokens;
    @SerializedName("authmanagerinfo") @Nullable private MwAuthManagerInfo amInfo;
    @Nullable private MarkReadResponse echomarkread;
    @Nullable private MarkReadResponse echomarkseen;
    @Nullable private NotificationList notifications;
    @Nullable private Map<String, Notification.UnreadNotificationWikiItem> unreadnotificationpages;
    @SerializedName("general") @Nullable private SiteInfo generalSiteInfo;
    @SerializedName("wikimediaeditortaskscounts") @Nullable private EditorTaskCounts editorTaskCounts;
    @Nullable private List<WatchlistItem> watchlist;
    @SerializedName("usercontribs") @Nullable private List<UserContribution> userContributions;

    @Nullable public List<MwQueryPage> pages() {
        return pages;
    }

    @Nullable public MwQueryPage firstPage() {
        if (pages != null && pages.size() > 0) {
            return pages.get(0);
        }
        return null;
    }

    @Nullable public UserInfo userInfo() {
        return userInfo;
    }

    @Nullable public String csrfToken() {
        return tokens != null ? tokens.csrf() : null;
    }

    @Nullable public String watchToken() {
        return  tokens != null ? tokens.watch() : null;
    }

    @Nullable public String createAccountToken() {
        return tokens != null ? tokens.createAccount() : null;
    }

    @Nullable public String loginToken() {
        return tokens != null ? tokens.login() : null;
    }

    @Nullable public NotificationList notifications() {
        return notifications;
    }

    @Nullable public Map<String, Notification.UnreadNotificationWikiItem> unreadNotificationWikis() {
        return unreadnotificationpages;
    }

    @Nullable public MarkReadResponse getEchoMarkSeen() {
        return echomarkseen;
    }

    @Nullable public String captchaId() {
        String captchaId = null;
        if (amInfo != null) {
            for (MwAuthManagerInfo.Request request : amInfo.requests()) {
                if ("CaptchaAuthenticationRequest".equals(request.id())) {
                    captchaId = request.fields().get("captchaId").value();
                }
            }
        }
        return captchaId;
    }

    @Nullable public ListUserResponse getUserResponse(@NonNull String userName) {
        if (users != null) {
            for (ListUserResponse user : users) {
                // MediaWiki user names are case sensitive, but the first letter is always capitalized.
                if (StringUtils.capitalize(userName).equals(user.name())) {
                    return user;
                }
            }
        }
        return null;
    }

    @NonNull public List<PageTitle> langLinks() {
        List<PageTitle> result = new ArrayList<>();
        if (pages == null || pages.isEmpty() || pages.get(0).langLinks() == null) {
            return result;
        }
        // noinspection ConstantConditions
        for (MwQueryPage.LangLink link : pages.get(0).langLinks()) {
            PageTitle title = new PageTitle(link.title(), WikiSite.forLanguageCode(link.lang()));
            result.add(title);
        }
        return result;
    }

    @Nullable public SiteInfo siteInfo() {
        return generalSiteInfo;
    }

    @Nullable public EditorTaskCounts editorTaskCounts() {
        return editorTaskCounts;
    }

    @NonNull public List<UserContribution> userContributions() {
        return userContributions != null ? userContributions : Collections.emptyList();
    }

    @NonNull public List<WatchlistItem> getWatchlist() {
        return watchlist != null ? watchlist : Collections.emptyList();
    }

    public boolean isEditProtected() {
        if (firstPage() == null || userInfo() == null) {
            return false;
        }
        for (Protection protection : firstPage().protection()) {
            if (protection.getType().equals("edit") && !userInfo().groups().contains(protection.getLevel())) {
                return true;
            }
        }
        return false;
    }

    @Override
    public void postProcess() {
        resolveConvertedTitles();
        resolveRedirectedTitles();
    }

    private void resolveRedirectedTitles() {
        if (redirects == null || pages == null) {
            return;
        }
        for (MwQueryPage page : pages) {
            for (MwQueryResult.Redirect redirect : redirects) {
                // TODO: Looks like result pages and redirects can also be matched on the "index"
                // property.  Confirm in the API docs and consider updating.
                if (page.title().equals(redirect.to())) {
                    page.redirectFrom(redirect.from());
                    if (redirect.toFragment() != null) {
                        page.appendTitleFragment(redirect.toFragment());
                    }
                }
            }
        }
    }

    private void resolveConvertedTitles() {
        if (converted == null || pages == null) {
            return;
        }
        for (MwQueryResult.ConvertedTitle convertedTitle : converted) {
            for (MwQueryPage page : pages) {
                if (page.title().equals(convertedTitle.to())) {
                    page.convertedFrom(convertedTitle.from());
                    page.convertedTo(convertedTitle.to());
                }
            }
        }
    }

    private static class Redirect {
        @SuppressWarnings("unused") private int index;
        @SuppressWarnings("unused") @Nullable private String from;
        @SuppressWarnings("unused") @Nullable private String to;
        @SuppressWarnings("unused") @SerializedName("tofragment") @Nullable private String toFragment;

        @Nullable public String to() {
            return to;
        }

        @Nullable public String from() {
            return from;
        }

        @Nullable public String toFragment() {
            return toFragment;
        }
    }

    public static class ConvertedTitle {
        @SuppressWarnings("unused") @Nullable private String from;
        @SuppressWarnings("unused") @Nullable private String to;

        @Nullable public String to() {
            return to;
        }

        @Nullable public String from() {
            return from;
        }
    }

    private static class Tokens {
        @SuppressWarnings("unused,NullableProblems") @SerializedName("csrftoken")
        @Nullable private String csrf;
        @SuppressWarnings("unused,NullableProblems") @SerializedName("createaccounttoken")
        @Nullable private String createAccount;
        @SuppressWarnings("unused,NullableProblems") @SerializedName("logintoken")
        @Nullable private String login;
        @SuppressWarnings("unused,NullableProblems") @SerializedName("watchtoken")
        @Nullable private String watch;

        @Nullable private String csrf() {
            return csrf;
        }
        @Nullable private String watch() {
            return watch;
        }

        @Nullable private String createAccount() {
            return createAccount;
        }

        @Nullable private String login() {
            return login;
        }
    }

    public static class MarkReadResponse {
        @SuppressWarnings("unused") @Nullable private String result;
        @SuppressWarnings("unused,NullableProblems") @Nullable private String timestamp;

        @Nullable public String getResult() {
            return result;
        }

        @Nullable public String getTimestamp() {
            return timestamp;
        }
    }

    public static class NotificationList {
        @SuppressWarnings("unused") private int count;
        @SuppressWarnings("unused") private int rawcount;
        @SuppressWarnings("unused") @Nullable private Notification.SeenTime seenTime;
        @SuppressWarnings("unused") @Nullable private List<Notification> list;
        @SuppressWarnings("unused") @SerializedName("continue") @Nullable private String continueStr;

        @Nullable public List<Notification> list() {
            return list;
        }

        @Nullable public String getContinue() {
            return continueStr;
        }

        public int getCount() {
            return count;
        }

        @Nullable public Notification.SeenTime getSeenTime() {
            return seenTime;
        }
    }

    public static class WatchlistItem {
        private int pageid;
        private long revid;
        @SerializedName("old_revid") private long oldRevid;
        private int ns;
        @Nullable private String title;
        @Nullable private String user;
        @Nullable private String timestamp;
        @Nullable private String comment;
        @Nullable private String parsedcomment;
        @Nullable private String logtype;
        private boolean anon;
        private boolean bot;
        @SerializedName("new") private boolean isNew;
        private boolean minor;
        private int oldlen;
        private int newlen;
        private WikiSite wiki;

        public int getNs() {
            return ns;
        }

        @NonNull public String getTitle() {
            return StringUtils.defaultString(title);
        }

        @NonNull public String getLogType() {
            return StringUtils.defaultString(logtype);
        }

        @NonNull public Date getDate() {
            return DateUtil.iso8601DateParse(StringUtils.defaultString(timestamp));
        }

        @NonNull public String getParsedComment() {
            return StringUtils.defaultString(parsedcomment);
        }

        public void setWiki(WikiSite wiki) {
            this.wiki = wiki;
        }

        public WikiSite getWiki() {
            return wiki;
        }

        @NonNull public String getUser() {
            return StringUtils.defaultString(user);
        }

        public int getOldlen() {
            return oldlen;
        }

        public int getNewlen() {
            return newlen;
        }

        public boolean isAnon() {
            return anon;
        }

        public long getRevid() {
            return revid;
        }
    }
}

File: app/src/main/java/org/wikipedia/dataclient/mwapi/MwQueryResult.kt
package org.wikipedia.dataclient.mwapi

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.decodeFromJsonElement
import org.wikipedia.dataclient.WikiSite
import org.wikipedia.json.JsonUtil
import org.wikipedia.notifications.db.Notification
import org.wikipedia.notifications.db.Notification.SeenTime
import org.wikipedia.notifications.db.Notification.UnreadNotificationWikiItem
import org.wikipedia.page.PageTitle
import org.wikipedia.settings.SiteInfo
import org.wikipedia.util.DateUtil
import org.wikipedia.util.StringUtil
import java.util.*

@Serializable
class MwQueryResult {

    @SerialName("userinfo") val userInfo: UserInfo? = null
    @SerialName("unreadnotificationpages") val unreadNotificationWikis: Map<String, UnreadNotificationWikiItem>? = null
    @SerialName("authmanagerinfo") private val amInfo: MwAuthManagerInfo? = null
    @SerialName("general") val siteInfo: SiteInfo? = null
    @SerialName("wikimediaeditortaskscounts") val editorTaskCounts: EditorTaskCounts? = null
    @SerialName("recentchanges") val recentChanges: List<RecentChange>? = null
    @SerialName("usercontribs") val userContributions: List<UserContribution> = emptyList()
    @SerialName("allusers") val allUsers: List<UserInfo>? = null
    @SerialName("globaluserinfo") val globalUserInfo: UserInfo? = null

    private val redirects: List<Redirect>? = null
    private val converted: List<ConvertedTitle>? = null
    private val tokens: Tokens? = null
    private val echomarkread: MarkReadResponse? = null
    val users: List<UserInfo>? = null
    val pages: List<MwQueryPage>? = null
    val echomarkseen: MarkReadResponse? = null
    val notifications: NotificationList? = null
    val watchlist: List<WatchlistItem> = emptyList()
    val namespaces: Map<String, Namespace>? = null
    val allmessages: List<Message>? = null
    val magicwords: List<MagicWord>? = null

    init {
        resolveConvertedTitles()
        resolveRedirectedTitles()
    }

    fun firstPage(): MwQueryPage? {
        return pages?.firstOrNull()
    }

    fun csrfToken(): String? {
        return tokens?.csrf
    }

    fun watchToken(): String? {
        return tokens?.watch
    }

    fun rollbackToken(): String? {
        return tokens?.rollback
    }

    fun createAccountToken(): String? {
        return tokens?.createAccount
    }

    fun loginToken(): String? {
        return tokens?.login
    }

    fun captchaId(): String? {
        val key = "captchaId"
        return amInfo?.requests?.find { it.fields?.containsKey(key) == true }?.fields?.get(key)?.value
    }

    fun getUserResponse(userName: String): UserInfo? {
        // MediaWiki user names are case sensitive, but the first letter is always capitalized.
        return users?.find { StringUtil.capitalize(userName) == it.name }
    }

    fun langLinks(): List<PageTitle> {
        val result = mutableListOf<PageTitle>()
        if (pages.isNullOrEmpty()) {
            return result
        }
        // noinspection ConstantConditions
        for (link in pages.first().langlinks) {
            val title = PageTitle(link.title, WikiSite.forLanguageCode(link.lang))
            result.add(title)
        }
        return result
    }

    val isEditProtected: Boolean
        get() {
            if (firstPage() == null || userInfo == null) {
                return false
            }
            for (protection in firstPage()!!.protection) {
                if (protection.type == "edit" && !userInfo.groups().contains(protection.level)) {
                    return true
                }
            }
            return false
        }

    private fun resolveRedirectedTitles() {
        if (redirects.isNullOrEmpty() || pages.isNullOrEmpty()) {
            return
        }
        for (page in pages) {
            for (redirect in redirects) {
                // TODO: Looks like result pages and redirects can also be matched on the "index"
                // property.  Confirm in the API docs and consider updating.
                if (page.title == redirect.to) {
                    page.redirectFrom = redirect.from
                    if (redirect.toFragment != null) {
                        page.appendTitleFragment(redirect.toFragment)
                    }
                }
            }
        }
    }

    private fun resolveConvertedTitles() {
        if (converted.isNullOrEmpty() || pages.isNullOrEmpty()) {
            return
        }
        converted.forEach { convertedTitle ->
            pages.filter { it.title == convertedTitle.to }.forEach { page ->
                page.convertedFrom = convertedTitle.from
                page.convertedTo = convertedTitle.to
            }
        }
    }

    @Serializable
    private class Redirect(@SerialName("tofragment") val toFragment: String? = null,
                           private val index: Int = 0,
                           val from: String? = null,
                           val to: String? = null)

    @Serializable
    class ConvertedTitle(val from: String? = null, val to: String? = null)

    @Serializable
    private class Tokens(@SerialName("csrftoken") val csrf: String? = null,
                         @SerialName("createaccounttoken") val createAccount: String? = null,
                         @SerialName("logintoken") val login: String? = null,
                         @SerialName("watchtoken") val watch: String? = null,
                         @SerialName("rollbacktoken") val rollback: String? = null)

    @Serializable
    class MarkReadResponse(val timestamp: String? = null, val result: String? = null)

    @Serializable
    class NotificationList(val list: List<Notification>? = null,
                           val seenTime: SeenTime? = null,
                           val count: Int = 0,
                           private val rawcount: Int = 0,
                           @SerialName("continue") val continueStr: String? = null)

    @Serializable
    class WatchlistItem {

        @SerialName("new") val isNew = false
        @SerialName("anon") val isAnon = false
        @SerialName("minor") val isMinor = false
        @SerialName("bot") val isBot = false
        @SerialName("old_revid") private val oldRevid: Long = 0
        private val timestamp: String? = null
        private val comment: String? = null
        val type: String = ""
        @SerialName("pageid") val pageId = 0
        val revid: Long = 0
        val ns = 0
        val title: String = ""
        val user: String = ""
        val logtype: String = ""
        val logdisplay: String = ""
        val oldlen = 0
        val newlen = 0
        var wiki: WikiSite? = null
        @SerialName("parsedcomment") val parsedComment: String = ""
        val date: Date
            get() = DateUtil.iso8601DateParse(timestamp.orEmpty())
    }

    @Serializable
    class RecentChange {
        private val type: String = ""
        private val ns = 0
        val title: String = ""
        private val pageid: Long = 0
        @SerialName("revid") val curRev: Long = 0
        @SerialName("old_revid") val revFrom: Long = 0
        val rcid: Long = 0
        val user: String = ""
        val anon = false
        val bot = false

        @SerialName("new") private val isNew = false
        private val minor = false
        val oldlen = 0
        val newlen = 0
        private val timestamp: String = ""

        @SerialName("parsedcomment") val parsedComment: String = ""
        private val tags: List<String>? = null
        private val oresscores: JsonElement? = null

        val parsedDateTime by lazy { DateUtil.iso8601LocalDateTimeParse(timestamp) }
        val joinedTags by lazy { tags?.joinToString(separator = ", ").orEmpty() }

        override fun toString(): String {
            return title
        }

        val ores: OresResult?
            get() = if (oresscores != null && oresscores !is JsonArray) {
                JsonUtil.json.decodeFromJsonElement<OresResult>(oresscores)
            } else null
    }

    @Serializable
    class OresResult {
        private val damaging: OresItem? = null
        private val goodfaith: OresItem? = null

        // TODO: articlequality
        // TODO: draftquality
        val damagingProb: Float
            get() = damaging?.trueProb ?: 0f
        val goodfaithProb: Float
            get() = goodfaith?.trueProb ?: 0f
    }

    @Serializable
    class OresItem {
        @SerialName("true") val trueProb = 0f
        @SerialName("false") val falseProb = 0f
    }

    @Serializable
    class Namespace {
        val id: Int = 0
        val name: String = ""
    }

    @Serializable
    class Message {
        val name: String = ""
        val content: String = ""
    }

    @Serializable
    class MagicWord {
        val name: String = ""
        val aliases: List<String> = emptyList()
    }
}
